### PR TITLE
Add support for enum to integer conversion

### PR DESCRIPTION
**Pull request by jccampanero**
Adds support for enum to integer conversion, according to `enum.ordinal()`. See #2963 for details.

@filiphr, for now I have left the work split into the commits of the original branch, so that code review can be done more easily and to make it easier to add additional changes if needed. I will squash the commits into one after the code review if you consider it appropriate.

Thank you very much.

### CODE REVIEW

File: processor/src/main/java/org/mapstruct/ap/internal/conversion/Conversions.java [hunk 1]
```diff
@@ -324,10 +330,14 @@ private void register(String sourceTypeName, Class<?> targetClass, ConversionPro
     }
 
     public ConversionProvider getConversion(Type sourceType, Type targetType) {
-        if ( sourceType.isEnumType() && targetType.equals( stringType ) ) {
+        if ( sourceType.isEnumType() &&
+                ( targetType.equals( stringType ) || targetType.equals( integerType ) || targetType.equals( intType ) )
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/conversion/Conversions.java:334]: You don't need the `intType` here you can do `targetType.getBoxedEquivalent().equals( integerType )` instead

File: processor/src/main/java/org/mapstruct/ap/internal/conversion/Conversions.java [hunk 2]
```diff
@@ -324,10 +330,14 @@ private void register(String sourceTypeName, Class<?> targetClass, ConversionPro
...
+        ) {
             sourceType = enumType;
         }
-        else if ( targetType.isEnumType() && sourceType.equals( stringType ) ) {
+        else if ( targetType.isEnumType() &&
+                ( sourceType.equals( stringType ) || sourceType.equals( integerType ) || sourceType.equals( intType ) )
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/conversion/Conversions.java:339]: You don't need the `intType` here you can do `sourceType.getBoxedEquivalent().equals( integerType )` instead

**Review by filiphr — CHANGES_REQUESTED** [review]
Looks great @jccampanero. There are only 2 tiny things that we should do before we merge this.

File: processor/src/main/java/org/mapstruct/ap/internal/conversion/Conversions.java (hunk 1 above)
Comment (jccampanero) [reply processor/src/main/java/org/mapstruct/ap/internal/conversion/Conversions.java]: Thank you very much for the feedback @filiphr. I implemented the requested change.

Please, let me know if any additional work needs to be done.

File: processor/src/main/java/org/mapstruct/ap/internal/conversion/Conversions.java (hunk 2 above)
Comment (jccampanero) [reply processor/src/main/java/org/mapstruct/ap/internal/conversion/Conversions.java]: Thank you very much again for the feedback @filiphr, I implemented the requested change.

Please do not hesitate to contact me to make any changes you deem appropriate.

File: processor/src/main/java/org/mapstruct/ap/internal/conversion/Conversions.java [hunk 3]
```diff
@@ -324,10 +328,18 @@ private void register(String sourceTypeName, Class<?> targetClass, ConversionPro
     }
 
     public ConversionProvider getConversion(Type sourceType, Type targetType) {
-        if ( sourceType.isEnumType() && targetType.equals( stringType ) ) {
+        if ( sourceType.isEnumType() &&
+                ( targetType.equals( stringType ) ||
+                  targetType.equals( integerType ) ||
+                  targetType.getBoxedEquivalent().equals( integerType ) )
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/conversion/Conversions.java:334]: When doing `getBoxedEquivalent` if the target type is not primitive it returns itself. This means that we can just do:

```suggestion
                  targetType.getBoxedEquivalent().equals( integerType ) )
```

File: processor/src/main/java/org/mapstruct/ap/internal/conversion/Conversions.java [hunk 4]
```diff
@@ -324,10 +328,18 @@ private void register(String sourceTypeName, Class<?> targetClass, ConversionPro
...
+        ) {
             sourceType = enumType;
         }
-        else if ( targetType.isEnumType() && sourceType.equals( stringType ) ) {
+        else if ( targetType.isEnumType() &&
+                ( sourceType.equals( stringType ) ||
+                  sourceType.equals( integerType ) ||
+                  sourceType.getBoxedEquivalent().equals( integerType ) )
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/conversion/Conversions.java:341]: When doing `getBoxedEquivalent` if the source type is not primitive it returns itself. This means that we can just do:

```suggestion
                  sourceType.getBoxedEquivalent().equals( integerType ) )
```

**Review by filiphr — CHANGES_REQUESTED** [review]
Thanks for the changes @jccampanero. One last request from my side and we are good to go to merge this.

**Comment by jccampanero** [comment]
Thank you very much @filiphr and sorry, I missed that point in my previous commit.

Please, let me know if you need anything else, I will gladly do it.

**Review by filiphr — APPROVED** [review]

**Comment by filiphr** [comment]
Thanks a lot for your contribution @jccampanero. I've merged this to main

**Comment by jccampanero** [comment]
You are welcome @filiphr; on the contrary, thank you for this great library.

**Pull request merged**